distributed/worker.py

```diff
@@ -1477,21 +1477,66 @@ def update_data(self, comm=None, data=None, report=True, serializers=None):
 
         if report:
+
+            self.log.append(
+                ("Notifying scheduler about in-memory in update-data", list(data))
+            )
             self.batched_stream.send({"op": "add-keys", "keys": list(data)})
         info = {"nbytes": {k: sizeof(v) for k, v in data.items()}, "status": "OK"}
         return info
 
-    def delete_data(self, comm=None, keys=None, report=True):
-        if keys:
-            for key in list(keys):
-                self.log.append((key, "delete"))
-                self.release_key(key, cause="delete data")
+    def free_keys_comm(self, comm, keys, reason):
+        """This is a comm handler calling `free_keys`"""
+        return self.handle_free_keys(keys, reason)
+
+    def handle_free_keys(self, keys, reason):
+        """
+        Stream handler to be called by the scheduler.
 
-            logger.debug("Worker %s -- Deleted %d keys", self.name, len(keys))
+        The given keys are no longer referred to and required by the scheduler.
+        The worker is now allowed to release the key, if applicable.
+
+        This does not guarantee that the memory is released since the worker may
+        still decide to hold on to the data and task since it is required by an
+        upstream dependency.
+        """
+        self.log.append(("free-keys", keys, reason))
+        for key in keys:
+            ts = self.tasks.get(key)
+            if ts:
+                ts.scheduler_holds_ref = False
+            self.release_key(key, report=False, reason=reason)
+
+    def handle_superfluous_data(self, keys=(), reason=None):
+        """Stream handler notifying the worker that it might be holding unreferenced, superfluous data.
+
+        This should not actually happen during ordinary operations and is only
+        intended to correct any erroneous state. An example where this is
+        necessary is if a worker fetches data for a downstream task but that
+        task is released before the data arrives.
+        In this case, the scheduler will notify the worker that it may be
+        holding this unnecessary data, if the worker hasn't released the data itself, already.
+
+        This handler does not guarantee the task nor the data to be actually
+        released but only asks the worker to release the data on a best effort
+        guarantee. This protects from race conditions where the given keys may
+        already have been rescheduled for compute in which case the compute
+        would win and this handler is ignored.
+
+        For stronger guarantees, see handler free_keys
+        """
+        self.log.append(("Handle superfluous data", keys, reason))
+        for key in list(keys):
+            ts = self.tasks.get(key)
+            if ts and not ts.scheduler_holds_ref:
+                self.release_key(key, reason=f"delete data: {reason}", report=False)
+
+        logger.debug("Worker %s -- Deleted %d keys", self.name, len(keys))
```

Comment:
I'm curious about why there should be multiple methods for this.  In principle I think of our current approach as the scheduler saying 

> I don't need you to have this data 

The worker though, as you say, may disagree and hold on to it.  I don't imagine a situation where the scheduler would override the worker, and so I don't see a reason to have multiple methods for this signal. 

Thoughts?

Reply:
First of all, I think this particular signal I called "superfluous data" should not exist and I believe this is what you're saying as well.
This signal is issued here (link to main branch) https://github.com/dask/distributed/blob/ddec3c9c7b5a5d3cb3b806ff25ee36ef640d8dda/distributed/scheduler.py#L6276-L6284

I am actually a bit fuzzy myself how this particular race condition occurs but something like this can happen but this is how I think it happens

```
Task A (Worker A) -->  Task C (Worker B)
           - ^
         /        
        /
Task B (Worker B)
```


* Worker B tries to execute C but needs to fetch Task A to do so
* Worker A submits data to B
* Data arrives at B and is being deserialized or is waiting for a loop, whatever
* Worker A dies and the scheduler notices
* Scheduler thinks all is lost and transitions Task A to released and tries to reschedule on Worker B
* ... In the meantime Worker B is done deserializing and puts key in memory, notifies scheduler usign `add-keys`
* Worker B receives the `compute-task` signal from scheduler and thinks he's lucky since he got the data already, cool. He tells the schdeuler -> `task-finished` signal
* Scheduler receives the add-keys response but thinks "This task is not in memory any more, please forget it" -> delete-data signal (what I call superfluous data)
* Worker B receives the `delete-data` signal and forgets *everything* (this is where my superfluous-data signal has a safeguard)
* Scheduler receives the `task-finished` signal from before and thinks everything is OK and proceeds as before, however, worker B does not have the data.


I assume this only is a problem due to another bug in how the worker corrects missing data (which I am also trying to fix) which would eventually heal the cluster but I'm not entirely certain

Reply:
Ha, that's a fun comedy of errors.

Yeah, so I agree that this feels like a different bug.  In principle it'd be nice to fix that bug rather than add this signal, but I can imagine that it might be best to do that in future work.